File: compilers/java/src/test/java/runar/compiler/IntentIntrinsicsTest.java
package runar.compiler;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;
import runar.compiler.frontend.GoParser;
import runar.compiler.ir.anf.AnfMethod;
import runar.compiler.ir.anf.AnfParam;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.AnfValue;
import runar.compiler.ir.anf.Call;
import runar.compiler.ir.ast.ContractNode;
import runar.compiler.passes.AnfLower;
import runar.compiler.passes.Typecheck;
import runar.compiler.passes.Validate;

/**
 * Tests for the intent sub-covenant intrinsics ({@code extractPrevOutputScript},
 * {@code requireOutputP2PKH}, {@code currentBlockHeight}) ported from the Go
 * tier (BSVM Phase 13). Mirrors
 * {@code compilers/go/frontend/intent_intrinsics_test.go}.
 */
class IntentIntrinsicsTest {

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static AnfProgram mustLower(String source) {
        ContractNode contract;
        try {
            contract = GoParser.parse(source, "Test.runar.go");
        } catch (GoParser.ParseException e) {
            throw new AssertionError("parse failed: " + e.getMessage(), e);
        }
        assertNotNull(contract, "parse returned null contract");
        Validate.run(contract);
        Typecheck.run(contract);
        return AnfLower.run(contract);
    }

    private static AnfMethod findMethod(AnfProgram p, String name) {
        for (AnfMethod m : p.methods()) {
            if (m.name().equals(name)) return m;
        }
        throw new AssertionError("method '" + name + "' not found in " + methodNames(p));
    }

    private static String methodNames(AnfProgram p) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < p.methods().size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append(p.methods().get(i).name());
        }
        return sb.append(']').toString();
    }

    private static List<String> paramNames(AnfMethod m) {
        List<String> out = new java.util.ArrayList<>(m.params().size());
        for (AnfParam p : m.params()) out.add(p.name());
        return out;
    }

    private static void expectTypeError(String source, String substr) {
        ContractNode contract;
        try {
            contract = GoParser.parse(source, "Test.runar.go");
        } catch (GoParser.ParseException e) {
            throw new AssertionError("parse failed: " + e.getMessage(), e);
        }
        assertNotNull(contract);
        Typecheck.TypeCheckException ex = assertThrows(
            Typecheck.TypeCheckException.class,
            () -> Typecheck.run(contract)
        );
        boolean matched = false;
        for (String e : ex.errors()) {
            if (e.contains(substr)) { matched = true; break; }
        }
        assertTrue(
            matched,
            "expected typecheck error containing '" + substr + "' but got: " + ex.errors()
        );
    }

    // ------------------------------------------------------------------
    // extractPrevOutputScript
    // ------------------------------------------------------------------

    @Test
    void extractPrevOutputScript_autoInjectsWitnessParam() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type IntentCov struct {
            \trunar.StatefulSmartContract
            \tStateCovScriptHash runar.ByteString `runar:"readonly"`
            }

            func (c *IntentCov) CoSpend() {
            \tstateCovScript := runar.ExtractPrevOutputScript(0, c.StateCovScriptHash)
            \t_ = stateCovScript
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "coSpend");
        List<String> names = paramNames(m);
        assertTrue(names.contains("_prevOutScript_0"),
            "expected param '_prevOutScript_0' in " + names);
        assertTrue(names.contains("txPreimage"),
            "expected param 'txPreimage' in " + names);
    }

    @Test
    void extractPrevOutputScript_twoIndicesProduceTwoParams() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type IntentCov struct {
            \trunar.StatefulSmartContract
            \tH0 runar.ByteString `runar:"readonly"`
            \tH1 runar.ByteString `runar:"readonly"`
            }

            func (c *IntentCov) CoSpend() {
            \ta := runar.ExtractPrevOutputScript(0, c.H0)
            \tb := runar.ExtractPrevOutputScript(1, c.H1)
            \t_ = a
            \t_ = b
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "coSpend");
        List<String> names = paramNames(m);
        assertTrue(names.contains("_prevOutScript_0"), "missing _prevOutScript_0 in " + names);
        assertTrue(names.contains("_prevOutScript_1"), "missing _prevOutScript_1 in " + names);
    }

    @Test
    void extractPrevOutputScript_sameIndexIsIdempotent() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type IntentCov struct {
            \trunar.StatefulSmartContract
            \tH0 runar.ByteString `runar:"readonly"`
            }

            func (c *IntentCov) CoSpend() {
            \ta := runar.ExtractPrevOutputScript(0, c.H0)
            \tb := runar.ExtractPrevOutputScript(0, c.H0)
            \t_ = a
            \t_ = b
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "coSpend");
        int count = 0;
        for (AnfParam prm : m.params()) {
            if ("_prevOutScript_0".equals(prm.name())) count++;
        }
        assertEquals(1, count,
            "expected exactly one _prevOutScript_0 param across duplicate calls");
    }

    @Test
    void extractPrevOutputScript_nonLiteralIndex_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type IntentCov struct {
            \trunar.StatefulSmartContract
            \tH0 runar.ByteString `runar:"readonly"`
            }

            func (c *IntentCov) CoSpend(idx runar.Bigint) {
            \t_ = runar.ExtractPrevOutputScript(idx, c.H0)
            }
            """;
        expectTypeError(src, "must be an integer literal");
    }

    // ------------------------------------------------------------------
    // requireOutputP2PKH
    // ------------------------------------------------------------------

    @Test
    void requireOutputP2PKH_autoInjectsSerialisedOutputs() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tBondPKH runar.ByteString `runar:"readonly"`
            \tBond    runar.Bigint     `runar:"readonly"`
            }

            func (c *Cov) PayBond() {
            \trunar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "payBond");
        assertTrue(paramNames(m).contains("_serialisedOutputs"),
            "expected '_serialisedOutputs' in " + paramNames(m));
    }

    @Test
    void requireOutputP2PKH_multipleCalls_oneSerialisedOutputsParam() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tBondPKH runar.ByteString `runar:"readonly"`
            \tBond    runar.Bigint     `runar:"readonly"`
            }

            func (c *Cov) PayMulti() {
            \trunar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
            \trunar.RequireOutputP2PKH(1, c.BondPKH, c.Bond)
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "payMulti");
        int count = 0;
        for (AnfParam prm : m.params()) {
            if ("_serialisedOutputs".equals(prm.name())) count++;
        }
        assertEquals(1, count,
            "expected exactly one _serialisedOutputs param across multiple intrinsic calls");
    }

    // ------------------------------------------------------------------
    // Crit-2 — extractPrevOutputScript prefix-hash 3-arg form
    // ------------------------------------------------------------------

    @Test
    void extractPrevOutputScript_prefixForm_lowersWithSubstr() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type IntentTemplate struct {
            \trunar.StatefulSmartContract
            \tExpectedPolicyPrefixHash runar.ByteString `runar:"readonly"`
            }

            func (c *IntentTemplate) Bind() {
            \ts := runar.ExtractPrevOutputScript(0, c.ExpectedPolicyPrefixHash, 600)
            \t_ = s
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "bind");
        // Expect a substr call inside the method body (the prefix
        // extraction preceding the hash256). Distinguish from any other
        // substr by checking it consumes a load_param ref to
        // _prevOutScript_0.
        boolean sawPrefixSubstr = false;
        var body = m.body();
        for (int i = 0; i < body.size(); i++) {
            AnfValue v = body.get(i).value();
            if (v instanceof Call c && "substr".equals(c.func()) && c.args().size() == 3) {
                String ref = c.args().get(0);
                for (int j = 0; j < i; j++) {
                    var prior = body.get(j);
                    if (prior.name().equals(ref)
                        && prior.value() instanceof runar.compiler.ir.anf.LoadParam lp
                        && "_prevOutScript_0".equals(lp.name())) {
                        sawPrefixSubstr = true;
                        break;
                    }
                }
                if (sawPrefixSubstr) break;
            }
        }
        assertTrue(sawPrefixSubstr,
            "expected substr(load_param(_prevOutScript_0), …) for 3-arg prefix form");
    }

    @Test
    void extractPrevOutputScript_prefixForm_nonLiteralPrefixLen_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tH runar.ByteString `runar:"readonly"`
            }

            func (c *Cov) Bind(n runar.Bigint) {
            \t_ = runar.ExtractPrevOutputScript(0, c.H, n)
            }
            """;
        expectTypeError(src, "prefixLen) must be an integer literal");
    }

    @Test
    void extractPrevOutputScript_tooManyArgs_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tH runar.ByteString `runar:"readonly"`
            }

            func (c *Cov) Bind() {
            \t_ = runar.ExtractPrevOutputScript(0, c.H, 600, 999)
            }
            """;
        expectTypeError(src, "expects 2 or 3 arguments");
    }

    // ------------------------------------------------------------------
    // Crit-3 — requireOutputP2PKH + addDataOutput mix rejection
    // ------------------------------------------------------------------

    @Test
    void requireOutputP2PKH_mixedWithAddDataOutput_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tBondPKH runar.ByteString `runar:"readonly"`
            \tBond    runar.Bigint     `runar:"readonly"`
            \tTag     runar.ByteString `runar:"readonly"`
            }

            func (c *Cov) PayBondAndAnnounce() {
            \tc.AddDataOutput(0, c.Tag)
            \trunar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
            }
            """;
        expectTypeError(src, "mixes requireOutputP2PKH() with addDataOutput()");
    }

    @Test
    void requireOutputP2PKH_withoutAddDataOutput_ok() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tBondPKH runar.ByteString `runar:"readonly"`
            \tBond    runar.Bigint     `runar:"readonly"`
            }

            func (c *Cov) PayBond() {
            \trunar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
            }
            """;
        mustLower(src); // must not throw
    }

    @Test
    void requireOutputP2PKH_nonLiteralIndex_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tBondPKH runar.ByteString `runar:"readonly"`
            \tBond    runar.Bigint     `runar:"readonly"`
            }

            func (c *Cov) PayBond(idx runar.Bigint) {
            \trunar.RequireOutputP2PKH(idx, c.BondPKH, c.Bond)
            }
            """;
        expectTypeError(src, "must be an integer literal");
    }

    // ------------------------------------------------------------------
    // currentBlockHeight
    // ------------------------------------------------------------------

    @Test
    void currentBlockHeight_desugarsToExtractLocktime() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Cov struct {
            \trunar.StatefulSmartContract
            \tDeadline runar.Bigint `runar:"readonly"`
            }

            func (c *Cov) Spend() {
            \th := runar.CurrentBlockHeight()
            \trunar.Assert(h <= c.Deadline)
            }
            """;
        AnfProgram p = mustLower(src);
        AnfMethod m = findMethod(p, "spend");
        boolean saw = false;
        for (var b : m.body()) {
            AnfValue v = b.value();
            if (v instanceof Call c && "extractLocktime".equals(c.func())) {
                saw = true;
                break;
            }
        }
        assertTrue(saw, "expected currentBlockHeight() to desugar to extractLocktime call");
    }

    @Test
    void currentBlockHeight_statelessContract_errors() {
        String src = """
            package x

            import runar "github.com/icellan/runar/packages/runar-go"

            type Sl struct {
            \trunar.SmartContract
            \tDeadline runar.Bigint `runar:"readonly"`
            }

            func (c *Sl) Spend() bool {
            \th := runar.CurrentBlockHeight()
            \treturn h > c.Deadline
            }
            """;
        expectTypeError(src, "StatefulSmartContract");
    }
}

File: compilers/go/frontend/intent_intrinsics_test.go
package frontend

import (
	"strings"
	"testing"

	"github.com/icellan/runar/compilers/go/ir"
)

// mustLowerGoSource parses, validates, typechecks, and ANF-lowers a Go-DSL
// source string. Returns the lowered program. Fails the test on any error.
func mustLowerGoSource(t *testing.T, source string) *ir.ANFProgram {
	t.Helper()
	result := ParseSource([]byte(source), "Test.runar.go")
	if len(result.Errors) > 0 {
		t.Fatalf("parse errors: %s", strings.Join(result.ErrorStrings(), "; "))
	}
	if result.Contract == nil {
		t.Fatal("parse returned nil contract")
	}
	valResult := Validate(result.Contract)
	if len(valResult.Errors) > 0 {
		t.Fatalf("validation errors: %s", strings.Join(valResult.ErrorStrings(), "; "))
	}
	tcResult := TypeCheck(result.Contract)
	if len(tcResult.Errors) > 0 {
		t.Fatalf("typecheck errors: %s", strings.Join(tcResult.ErrorStrings(), "; "))
	}
	return LowerToANF(result.Contract)
}

// expectIntrinsicTypeError asserts that the source produces a typecheck
// error containing `substr`. Used for negative tests (e.g. non-literal
// indices, stateless misuse).
func expectIntrinsicTypeError(t *testing.T, source string, substr string) {
	t.Helper()
	result := ParseSource([]byte(source), "Test.runar.go")
	if len(result.Errors) > 0 {
		t.Fatalf("unexpected parse errors: %s", strings.Join(result.ErrorStrings(), "; "))
	}
	tcResult := TypeCheck(result.Contract)
	for _, d := range tcResult.Errors {
		if strings.Contains(d.FormatMessage(), substr) {
			return
		}
	}
	var msgs []string
	for _, d := range tcResult.Errors {
		msgs = append(msgs, d.FormatMessage())
	}
	t.Fatalf("expected typecheck error containing %q, got: %v", substr, msgs)
}

// findANFMethod returns the named method from a lowered program.
func findANFMethod(t *testing.T, p *ir.ANFProgram, name string) *ir.ANFMethod {
	t.Helper()
	for i := range p.Methods {
		if p.Methods[i].Name == name {
			return &p.Methods[i]
		}
	}
	t.Fatalf("method %q not found; got: %v", name, methodNames(p))
	return nil
}

func methodNames(p *ir.ANFProgram) []string {
	out := make([]string, len(p.Methods))
	for i := range p.Methods {
		out[i] = p.Methods[i].Name
	}
	return out
}

func paramNamesOf(m *ir.ANFMethod) []string {
	out := make([]string, len(m.Params))
	for i := range m.Params {
		out[i] = m.Params[i].Name
	}
	return out
}

// ---------------------------------------------------------------------------
// extractPrevOutputScript
// ---------------------------------------------------------------------------

func TestExtractPrevOutputScript_AutoInjectsWitnessParam(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type IntentCov struct {
	runar.StatefulSmartContract
	StateCovScriptHash runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *IntentCov) CoSpend() {
	stateCovScript := runar.ExtractPrevOutputScript(0, c.StateCovScriptHash)
	_ = stateCovScript
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "coSpend")
	names := paramNamesOf(m)
	want := map[string]bool{"_prevOutScript_0": false, "txPreimage": false}
	for _, n := range names {
		if _, ok := want[n]; ok {
			want[n] = true
		}
	}
	for name, found := range want {
		if !found {
			t.Errorf("expected auto-injected param %q in %s.Params, got %v", name, m.Name, names)
		}
	}
}

func TestExtractPrevOutputScript_TwoIndicesProduceTwoParams(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type IntentCov struct {
	runar.StatefulSmartContract
	H0 runar.ByteString ` + "`runar:\"readonly\"`" + `
	H1 runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *IntentCov) CoSpend() {
	a := runar.ExtractPrevOutputScript(0, c.H0)
	b := runar.ExtractPrevOutputScript(1, c.H1)
	_ = a
	_ = b
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "coSpend")
	names := paramNamesOf(m)
	for _, want := range []string{"_prevOutScript_0", "_prevOutScript_1"} {
		found := false
		for _, n := range names {
			if n == want {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("expected auto-injected param %q in %s.Params, got %v", want, m.Name, names)
		}
	}
}

func TestExtractPrevOutputScript_SameIndexIsIdempotent(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type IntentCov struct {
	runar.StatefulSmartContract
	H0 runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *IntentCov) CoSpend() {
	a := runar.ExtractPrevOutputScript(0, c.H0)
	b := runar.ExtractPrevOutputScript(0, c.H0)
	_ = a
	_ = b
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "coSpend")
	count := 0
	for _, prm := range m.Params {
		if prm.Name == "_prevOutScript_0" {
			count++
		}
	}
	if count != 1 {
		t.Errorf("expected exactly one _prevOutScript_0 param, got %d", count)
	}
}

// R-2 / R-4 — typecheck bounds checks ------------------------------------

func TestRequireOutputP2PKH_OutputIndexBound_Rejects(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	PKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	A   runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Pay() {
	// 2000 > 1000 bound — should be rejected at typecheck.
	runar.RequireOutputP2PKH(2000, c.PKH, c.A)
}
`
	expectIntrinsicTypeError(t, source, "bound to <= 1000")
}

func TestRequireOutputP2PKH_NegativeIndex_Rejects(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	PKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	A   runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Pay() {
	runar.RequireOutputP2PKH(-1, c.PKH, c.A)
}
`
	expectIntrinsicTypeError(t, source, "must be >= 0")
}

func TestExtractPrevOutputScript_PrefixLenTooSmall_Rejects(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	H runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Bind() {
	// prefixLen=16 < 32 (hash size) — should be rejected.
	_ = runar.ExtractPrevOutputScript(0, c.H, 16)
}
`
	expectIntrinsicTypeError(t, source, "must be >= 32")
}

func TestExtractPrevOutputScript_PrefixLenTooLarge_Rejects(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	H runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Bind() {
	// prefixLen=10485760 > 4 MiB — should be rejected.
	_ = runar.ExtractPrevOutputScript(0, c.H, 10485760)
}
`
	expectIntrinsicTypeError(t, source, "MAX_SCRIPT_BYTES")
}

// Crit-2 — prefix-hash 3-arg form ----------------------------------------

func TestExtractPrevOutputScript_PrefixForm_LowersWithSubstr(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type IntentTemplate struct {
	runar.StatefulSmartContract
	ExpectedPolicyPrefixHash runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *IntentTemplate) Bind() {
	s := runar.ExtractPrevOutputScript(0, c.ExpectedPolicyPrefixHash, 600)
	_ = s
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "bind")
	// Expect a substr call inside the method body (the prefix extraction
	// preceding the hash256). Distinguish from any other substr by
	// checking it consumes a load_param ref + a 0 literal + a 600 literal.
	sawPrefixSubstr := false
	for i, b := range m.Body {
		if b.Value.Kind == "call" && b.Value.Func == "substr" && len(b.Value.Args) == 3 {
			// The first arg should be a load_param for _prevOutScript_0
			ref := b.Value.Args[0]
			for j := 0; j < i; j++ {
				if m.Body[j].Name == ref &&
					m.Body[j].Value.Kind == "load_param" &&
					m.Body[j].Value.Name == "_prevOutScript_0" {
					sawPrefixSubstr = true
					break
				}
			}
			if sawPrefixSubstr {
				break
			}
		}
	}
	if !sawPrefixSubstr {
		t.Errorf("expected substr(load_param(_prevOutScript_0), …) for 3-arg prefix form; body=%v", m.Body)
	}
}

func TestExtractPrevOutputScript_PrefixForm_NonLiteralPrefixLen_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	H runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Bind(n runar.Bigint) {
	_ = runar.ExtractPrevOutputScript(0, c.H, n)
}
`
	expectIntrinsicTypeError(t, source, "prefixLen) must be an integer literal")
}

func TestExtractPrevOutputScript_TooManyArgs_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	H runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Bind() {
	_ = runar.ExtractPrevOutputScript(0, c.H, 600, 999)
}
`
	expectIntrinsicTypeError(t, source, "expects 2 or 3 arguments")
}

// Crit-3 — addDataOutput + requireOutputP2PKH mix rejection ----------------

func TestRequireOutputP2PKH_MixedWithAddDataOutput_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	BondPKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	Bond    runar.Bigint     ` + "`runar:\"readonly\"`" + `
	Tag     runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) PayBondAndAnnounce() {
	c.AddDataOutput(0, c.Tag)
	runar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
}
`
	expectIntrinsicTypeError(t, source, "mixes requireOutputP2PKH() with addDataOutput()")
}

func TestRequireOutputP2PKH_WithoutAddDataOutput_OK(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	BondPKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	Bond    runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) PayBond() {
	runar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
}
`
	mustLowerGoSource(t, source) // must not error
}

// -------------------------------------------------------------------------

func TestExtractPrevOutputScript_NonLiteralIndex_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type IntentCov struct {
	runar.StatefulSmartContract
	H0 runar.ByteString ` + "`runar:\"readonly\"`" + `
}

func (c *IntentCov) CoSpend(idx runar.Bigint) {
	_ = runar.ExtractPrevOutputScript(idx, c.H0)
}
`
	expectIntrinsicTypeError(t, source, "must be an integer literal")
}

// ---------------------------------------------------------------------------
// requireOutputP2PKH
// ---------------------------------------------------------------------------

func TestRequireOutputP2PKH_AutoInjectsSerialisedOutputs(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	BondPKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	Bond    runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) PayBond() {
	runar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "payBond")
	names := paramNamesOf(m)
	found := false
	for _, n := range names {
		if n == "_serialisedOutputs" {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("expected auto-injected param '_serialisedOutputs' in %s.Params, got %v", m.Name, names)
	}
}

func TestRequireOutputP2PKH_MultipleCalls_OneSerialisedOutputsParam(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	BondPKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	Bond    runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) PayMulti() {
	runar.RequireOutputP2PKH(0, c.BondPKH, c.Bond)
	runar.RequireOutputP2PKH(1, c.BondPKH, c.Bond)
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "payMulti")
	count := 0
	for _, prm := range m.Params {
		if prm.Name == "_serialisedOutputs" {
			count++
		}
	}
	if count != 1 {
		t.Errorf("expected exactly one _serialisedOutputs param across multiple intrinsic calls, got %d", count)
	}
}

func TestRequireOutputP2PKH_NonLiteralIndex_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	BondPKH runar.ByteString ` + "`runar:\"readonly\"`" + `
	Bond    runar.Bigint     ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) PayBond(idx runar.Bigint) {
	runar.RequireOutputP2PKH(idx, c.BondPKH, c.Bond)
}
`
	expectIntrinsicTypeError(t, source, "must be an integer literal")
}

// ---------------------------------------------------------------------------
// currentBlockHeight
// ---------------------------------------------------------------------------

func TestCurrentBlockHeight_DesugarsToExtractLocktime(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	Deadline runar.Bigint ` + "`runar:\"readonly\"`" + `
}

func (c *Cov) Spend() {
	h := runar.CurrentBlockHeight()
	runar.Assert(h <= c.Deadline)
}
`
	p := mustLowerGoSource(t, source)
	m := findANFMethod(t, p, "spend")
	// Expect the body to contain at least one ANF call to extractLocktime
	// and a load_param("txPreimage") feeding it.
	sawExtractLocktime := false
	for _, b := range m.Body {
		if b.Value.Kind == "call" && b.Value.Func == "extractLocktime" {
			sawExtractLocktime = true
			break
		}
	}
	if !sawExtractLocktime {
		t.Errorf("expected currentBlockHeight() to desugar to extractLocktime call in %s.Body", m.Name)
	}
}

// TestAffineChecker_LenBranchWithStateMutations validates the hand-off §3
// concern that `if runar.Len(x) > 0 { state-mutation } else { state-mutation }`
// — branching on a read-only intrinsic value with state mutations on both
// arms — passes the affine type checker. This is the precondition for
// folding AdvanceStatePrivileged into AdvanceState (BSVM Phase 13 §3).
// Regression gate.
func TestAffineChecker_LenBranchWithStateMutations(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Cov struct {
	runar.StatefulSmartContract
	Count runar.Bigint
	Tag   runar.ByteString
}

func (c *Cov) Maybe(scratch runar.ByteString) {
	if runar.Len(scratch) > 0 {
		c.Count = c.Count + 1
		c.Tag = scratch
	} else {
		c.Count = c.Count - 1
		c.Tag = runar.ByteString("00")
	}
	c.AddOutput(1000, c.Count, c.Tag)
}
`
	// mustLowerGoSource fails the test on any typecheck error, so reaching
	// this point means the affine checker accepted the branched body.
	p := mustLowerGoSource(t, source)
	if findANFMethod(t, p, "maybe") == nil {
		t.Fatal("expected method 'maybe' in lowered program")
	}
}

func TestCurrentBlockHeight_StatelessContract_Errors(t *testing.T) {
	source := `
package x

import runar "github.com/icellan/runar/packages/runar-go"

type Sl struct {
	runar.SmartContract
	Deadline runar.Bigint ` + "`runar:\"readonly\"`" + `
}

func (c *Sl) Spend() bool {
	h := runar.CurrentBlockHeight()
	return h > c.Deadline
}
`
	expectIntrinsicTypeError(t, source, "StatefulSmartContract")
}
